fix(amplify-category-function): add length validation for secret value

SebasRod23 opened the pull request:
#### Description of changes
Validate that the secret value provided by the user has a length between 1 and 2048 characters

#### Issue #, if available
Closes #8304

#### Description of how you validated changes
The cli now doesn't allow you to continue until you have provided a valid secret value, similar to how the validation of the secret name works.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

SebasRod23 committed:
fix(amplify-category-function): add length validation for secret value

Validate that the secret value provided by the user has a length between 1 and 2048 characters

Closes #8304

danielleadams reviewed (changes requested):
Hi @SebasRod23, thanks for the contribution. Do you mind adding a unit test? You can add a new file named `secretValuesWalkthrough.test.ts` to the `__tests__/provider-utils/awscloudformation/service-walkthroughs` directory.

SebasRod23 commented:
No problem, I can do that

SebasRod23 committed:
test: add unit test for secret value walkthrough

Add unit tests for the secret value validation function

cjihrig commented on packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
 const secretValueDefaultMessage = (secretName: string) => `Enter the value for ${secretName}:`;
 
+export const secretValueValidator = () => (input?: string) => {
+  if (!input || input.length === 0 || input.length > 2048) {
```
Are the `!input` and `input.length === 0` checks both necessary?

SebasRod23 replied:
I think `input.length === 0` should be enough, but I added `!input` just in case `input` got a value of `undefined`

cjihrig replied:
Instead of `!input`, can you use `typeof input !== 'string'`.

SebasRod23 replied:
Sure, no problem

cjihrig reviewed (commented):
LGTM, with a small question.

SebasRod23 committed:
refactor: check that the input value is of type string

Change one of the conditions of the validation to check that the input value provided is of type
string

danielleadams commented on packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
       type: 'password',
       name: 'secretValue',
       message,
+      validate: secretValueValidator(),
```
This looks good - I don't think you need to invoked `secretValueValidator` here. I think you can pass in the function without invoking it and removing the top level function from `secretValueValidator`.

danielleadams commented on packages/amplify-category-function/src/provider-utils/awscloudformation/service-walkthroughs/secretValuesWalkthrough.ts:
```diff
@@ -176,12 +176,20 @@ const enterSecretName = async (invalidNames: string[]) =>
 
 const secretValueDefaultMessage = (secretName: string) => `Enter the value for ${secretName}:`;
 
+export const secretValueValidator = () => (input?: string) => {
```
I think the `validate` parameter only needs the function, and doesn't need the top level one.
```suggestion
export const secretValueValidator = (input?: string) => {
```

SebasRod23 replied:
Sure, that seems better

SebasRod23 committed:
refactor: update validate function for secret value

khrisrichardson commented:
Per the [SecretsManager documentation](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_CreateSecret.html#API_CreateSecret_RequestParameters), the length constraints for both SecretBinary and SecretString are

```
Length Constraints: Minimum length of 0. Maximum length of 65536.
```

Per the [Systemsmanager documentation](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#API_PutParameter_RequestParameters), the length constraints for parameter values are as follows

```
Standard parameters have a value limit of 4 KB. Advanced parameters have a value limit of 8 KB.
```

In other words the limit of 2048 characters is too constrained.
